### ISSUE TITLE
Cascaded Shadow Generator

### ISSUE THREAD

**Issue by dennemark** [brianzinn/react-babylonjs#83]
Hi,
thanks a lot for setting up this project, so far it is very useful!
I am missing the Cascaded Shadow Generator. Is it possible to integrate it? I tried to just add it to a directional light, however castShadows attribute of course wont work with it. I guess it should be pretty similar to Shadow Generator.
Furthermore, I would be interested, how this will be implemented, so I hope I can support this project in the future. 
Best!

**Comment by brianzinn** [comment]
That's very cool - I didn't even know about that one :)  I see it was added earlier this year.  The shadow generator code is all generated, so I will need to update the generator (https://github.com/brianzinn/react-babylonjs/blob/master/tools/generate-code.ts#L142).  I don't think it will take much time to implement, but it does have a different lifecycle management and has a deferred creation.  I am super busy this week, but maybe on Thursday if I can get some time then.  The other trick with shadow generators is that their construction is deferred until the light is created - this staged creation is handled automatically.  Thanks for using the project and I am excited to try it out.  I just checked the docs as well, looks like `splitFrustum()` is called automatically unless you change camera near/far plane, so should be ok to work declaratively the same as the current one!  (Also, I'm not sure about multiple shadow generators like the doc playground...).  Does your project with with Shadow Generator, but you just want to swap over to Cascaded Shadow Generator?

**Comment by dennemark** [comment]
I just created a pull request for cascaded shadow maps (CSM) #84 . Managed to figure out how to generate the code myself as the call of the script is mentioned in the package.json
So far it seems to work. Concerning multiple shadow generators I only need the CSM so far.
Only **issue** I still have is assigning shadow casters. I thought castShadows as attribute would lead to shadow casters, but so far I wasnt successful with it. Is this the correct way to create shadows? Maybe we should better test this before accepting the pull request.

**Comment by dennemark** [comment]
Just managed to create shadow caster for a simple sphere, so it seems to work. However it does not seem to work for a <Model>
I thought casting shadows would be done by writing the attribute castShadows on the sphere, since I read it somewhere in the issues. shadowCasters seems to be the right way to go. I guess placing a ref onto cascadedShadowGenerator allows me to add casters programmatically. I.e. i want to instanciate non-declaratively trees that should cast shadows, I guess I need to assign them individually to the caster list through a ref or something else.

`<sphere   name="sphere1" />
<directionalLight>
              <cascadedShadowGenerator shadowCasters={["sphere1"]}></cascadedShadowGenerator>
 </directionalLight>`

**Comment by brianzinn** [comment]
yes, if you get the ref then you can add shadow casters yourself:
`https://doc.babylonjs.com/api/classes/babylon.cascadedshadowgenerator#addshadowcaster`

that's what the lifecycle listener does automatically by mesh name, but you can add meshes yourself:
```
const shadowCasterRef = useRef(null);

// make sure this isn't called every render :)
if (shadowCasterRef.current) {
   const newMesh = createMesh();
   shadowCasterRef.current.hostInstance.addShadowCaster(newMesh);
}

return (
   <cascadedShadowCaster ref={shadowCasterRef} ... />
)
```

**Comment by brianzinn** [comment]
I've published in v2.2.5!  Thank you so much for your contribution.  Please re-open if you have any questions or related feature requests.
